core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/runtimeModels.scala

```diff
@@ -0,0 +1,344 @@
+package org.broadinstitute.dsde.workbench.leonardo
+
+import java.net.URL
+import java.time.Instant
+import java.util.UUID
+
+import cats.implicits._
+import enumeratum.{Enum, EnumEntry}
+import org.broadinstitute.dsde.workbench.google2.MachineTypeName
+import org.broadinstitute.dsde.workbench.leonardo.DataprocRole.SecondaryWorker
+import org.broadinstitute.dsde.workbench.leonardo.RuntimeContainerServiceType.JupyterService
+import org.broadinstitute.dsde.workbench.leonardo.RuntimeImageType.{CustomDataProc, Jupyter, RStudio, Welder}
+import org.broadinstitute.dsde.workbench.model.google.{parseGcsPath, GcsBucketName, GcsPath, GoogleProject}
+import org.broadinstitute.dsde.workbench.model.{ValueObject, WorkbenchEmail}
+
+import scala.collection.immutable
+
+/**
+ * This file contains models for Leonardo runtimes.
+ */
+/** The runtime itself */
+final case class Runtime(id: Long = 0, // DB AutoInc
+                         internalId: RuntimeInternalId,
+                         runtimeName: RuntimeName,
+                         googleProject: GoogleProject,
+                         serviceAccountInfo: ServiceAccountInfo,
+                         asyncRuntimeFields: Option[AsyncRuntimeFields],
+                         auditInfo: AuditInfo,
+                         dataprocProperties: Map[String, String],
+                         proxyUrl: URL,
+                         status: RuntimeStatus,
+                         labels: LabelMap,
+                         jupyterExtensionUri: Option[GcsPath],
+                         jupyterUserScriptUri: Option[UserScriptPath],
+                         jupyterStartUserScriptUri: Option[UserScriptPath],
+                         errors: List[RuntimeError],
+                         dataprocInstances: Set[DataprocInstance],
+                         userJupyterExtensionConfig: Option[UserJupyterExtensionConfig],
+                         autopauseThreshold: Int,
+                         defaultClientId: Option[String],
+                         stopAfterCreation: Boolean,
+                         allowStop: Boolean,
+                         runtimeImages: Set[RuntimeImage],
+                         scopes: Set[String],
+                         welderEnabled: Boolean,
+                         customClusterEnvironmentVariables: Map[String, String],
+                         runtimeConfigId: RuntimeConfigId) {
+  def projectNameString: String = s"${googleProject.value}/${runtimeName.asString}"
+  def nonPreemptibleInstances: Set[DataprocInstance] = dataprocInstances.filterNot(_.dataprocRole == SecondaryWorker)
+}
+
+object Runtime {
+  def addAsyncFields(runtime: Runtime, operation: Operation, stagingBucket: GcsBucketName): Runtime =
```

Comment:
maybe add `Lens[Runtime, AsyncRuntimeFields]` in https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/LeoLenses.scala?

```
  val asyncRuntimeFieldsRuntime: Lens[Runtime, AsyncRuntimeFields] = GenLens[Runtime](x =>
    AsyncRuntimeFields(???)
  )
```
And use
 `asyncRuntimeFieldsRuntime.set(asyncFields)(originalRuntime)` for updating fields

Reply:
Huh this method is never actually used… I added a lens anyway